storage/innobase/include/fil0fil.h

```diff
@@ -1404,19 +1398,32 @@ struct fil_system_t
   /** Map of fil_space_t::id to fil_space_t* */
   hash_table_t spaces;
 
-  /** whether each write to data files is durable (O_DSYNC) */
+  /** false=invoke fsync() or fdatasync() on data files before checkpoint;
+  true=each write is durable (O_DSYNC) */
   my_bool write_through;
   /** whether data files are buffered (not O_DIRECT) */
   my_bool buffered;
+  /** whether fdatasync() is needed on data files */
+  Atomic_relaxed<bool> need_unflushed_spaces;
 
   /** Try to enable or disable write-through of data files */
   void set_write_through(bool write_through);
+  /** Update innodb_doublewrite */
+  void set_use_doublewrite(ulong use)
+  {
+    buf_dblwr.set_use(use);
+    need_unflushed_spaces= !write_through && buf_dblwr.need_fsync();
```

Comment:
should "buffered" be checked for "need_unflushed_spaces"? buffered implies need_unflushed_spaces = true, regardless  buf_dblwr.need_fsync() ?

Reply:
I don’t think that `fil_system.buffered` (a.k.a. `innodb_data_file_buffering`) should be checked here. I believe that also writes via `O_DIRECT` may require `fdatasync` or `fsync` to be invoked, in case the file had been previously extended with a special system call, be it `fallocate` to allocate physical blocks or `ftruncate` to set the logical file size. In both cases, some recent writes may have impacted the file system metadata, that is, associate the written block with the inode so that a `read` like system call will return data from that block, instead of pretending that the region is filled with NUL bytes. If we always extended files via `pwrite` or `WriteFile` and invoked `fdatasync` or similar after such file-extending calls, then yes, we might safely skip `fdatasync` for subsequent overwrite-in-place. But even then, I suppose that invoking `fdatasync` would have the benefit that it would force a write-back of any non-durable cache inside the storage layer.